balancer/pickfirst/pickfirstleaf/pickfirstleaf.go

```diff
@@ -664,12 +698,69 @@ func (b *pickfirstBalancer) endFirstPassIfPossibleLocked(lastErr error) {
 	// Start re-connecting all the SubConns that are already in IDLE.
 	for _, v := range b.subConns.Values() {
 		sd := v.(*scData)
-		if sd.state == connectivity.Idle {
+		if sd.connectivityState == connectivity.Idle {
 			sd.subConn.Connect()
 		}
 	}
 }
 
+func (b *pickfirstBalancer) isActiveSCData(sd *scData) bool {
+	activeSD, found := b.subConns.Get(sd.addr)
+	return found && activeSD == sd
+}
+
+func (b *pickfirstBalancer) updateSubConnHealthState(sd *scData, state balancer.SubConnState) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	// Previously relevant SubConns can still callback with state updates.
+	// To prevent pickers from returning these obsolete SubConns, this logic
+	// is included to check if the current list of active SubConns includes
+	// this SubConn.
+	if !b.isActiveSCData(sd) {
+		return
+	}
+	switch state.ConnectivityState {
+	case connectivity.Ready:
+		b.updateConcludedStateLocked(balancer.State{
+			ConnectivityState: connectivity.Ready,
+			Picker:            &picker{result: balancer.PickResult{SubConn: sd.subConn}},
+		})
+	case connectivity.TransientFailure:
+		b.updateConcludedStateLocked(balancer.State{
+			ConnectivityState: connectivity.TransientFailure,
+			Picker:            &picker{err: fmt.Errorf("pickfirst: health check failure: %v", state.ConnectionError)},
+		})
+	case connectivity.Connecting:
+		b.updateConcludedStateLocked(balancer.State{
+			ConnectivityState: connectivity.Connecting,
+			Picker:            &picker{err: balancer.ErrNoSubConnAvailable},
+		})
+	default:
+		b.logger.Errorf("Got unexpected health update for SubConn %p: %v", state)
+	}
+}
+
+// updateConcludedStateLocked stores the state reported to the channel and calls
+// ClientConn.UpdateState(). As an optimization, it avoid sending duplicate
+// updates to the channel for state CONNECTING.
+func (b *pickfirstBalancer) updateConcludedStateLocked(newState balancer.State) {
+	// Optimization to not send duplicate CONNECTING updates.
+	if newState.ConnectivityState == b.concludedState && b.concludedState == connectivity.Connecting {
+		return
+	}
+	b.forceUpdateConcludedStateLocked(newState)
+}
+
+// forceUpdateConcludedStateLocked stores the state reported to the channel and
+// calls ClientConn.UpdateState().
+// A separate function is defined to force update the ClientConn state since the
+// channel doesn't correctly assume that LB policies start in CONNECTING and
+// relies on LB policy to send an initial CONNECTING update.
```

Comment:
Should we change that assumption in the channel?  File an issue and add a TODO to clean this up?  Maybe it's related to #7686 in some way?  I believe the assumption in the other languages is that LB policies start CONNECTING, and queue picks.

Reply:
This is related to https://github.com/grpc/grpc-go/issues/7686. I can do that in a separate PR. I was thinking of sending a picker and state update in the `ccBalancerWrapper` constructor here
https://github.com/grpc/grpc-go/blob/dcba136b362e8a8096b3986e700047f8ca6302ac/balancer_wrapper.go#L75-L94

Is this the correct place to make this change?